### PR TITLE
Added new metrics for record encyption

### LINKED ISSUE #789: Minimal Envelope encryption metrics
It would be good to have some counters to expose what is happening in the Encryption Filter:

1. How many records the Filter has encrypted/decrypted/left-unmodified
2. Total record bytes read by the filter and written by the filter (to compare the size of the records in/out) edit: perhaps limit to the encrypted records case? if we recorded all records it might be polluted by records we choose not to encrypt?
3. How many DEKs the proxy has generated/decrypted
4. How many Exceptions have been encountered when attempting to generate/decrypte DEKs
5. Count of other Exceptions thrown during encrypt/decrypt
6. How many times we have exhausted/rotated DEKs

Also it would be good to have some visibility of what's happening in the Vault KMS

1. How many HTTP requests have been made to Vault
2. HTTP Response counts by status code
3. Exception counts

edit: for now I have put off defining latency/performance related metrics as they feel like a lower priority (still going to be needed by the time we're happy it's production ready though)

### PR DESCRIPTION
**Pull request by ShubhamRwt**
### Type of change

_Select the type of your PR_

- Bugfix
- Enhancement / new feature
- Refactoring
- Documentation

### Description

This PR fixes #789. This PR add the new metrics `kroxylicious_encrypted_records_total` and `"kroxylicious_encrypted_records_total` which counts the records that are encrypted/not encyrpted.

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] PR raised from a fork of this repository and made from a branch rather than main. 
- [ ] Write tests
- [ ] Update documentation
- [ ] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [ ] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

### CODE REVIEW

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 1]
```diff
@@ -139,6 +145,17 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
                 });
     }
 
+    private void generateMetrics(String virtualClusterName, Set<String> unresolvedTopicNames, Map<String, TopicProduceData> topicNameToData) {
```
Comment (ShubhamRwt) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: This is how I am currently to counting the record for the topics that are not encrypted and creating individual counters for the topics

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java [hunk 2]
```diff
@@ -147,6 +156,10 @@ private CompletionStage<MemoryRecords> attemptEncrypt(@NonNull String topicName,
         });
     }
 
+    private void initMetrics(String virtualClusterName, String topicName) {
```
Comment (ShubhamRwt) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java]: This is not used method, Currently thinking if I can use this instead of initialising the metrics directly in the above method

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                                                                                                                                                         .description(
+                                                                                                                                                                 "Incremented by the number of records encrypted.")
+                                                                                                                                                         .tag(VIRTUAL_CLUSTER_LABEL,
+                                                                                                                                                                 virtualClusterName)
+                                                                                                                                                         .tag(TOPIC_NAME,
+                                                                                                                                                                 nodeId)
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: nodeId?

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java [hunk 4]
```diff
@@ -102,13 +104,18 @@ private ByteBufferOutputStream allocateBufferForEncrypt(@NonNull MemoryRecords r
...
-    private CompletionStage<MemoryRecords> attemptEncrypt(@NonNull String topicName,
+    private CompletionStage<MemoryRecords> attemptEncrypt(@NonNull String virtualClusterName,
+                                                          @NonNull String topicName,
                                                           int partition,
                                                           @NonNull EncryptionScheme<K> encryptionScheme,
                                                           @NonNull MemoryRecords records,
                                                           int attempt,
                                                           @NonNull IntFunction<ByteBufferOutputStream> bufferAllocator,
                                                           int allRecordsCount) {
+        var encyptedRecordsTotal = RecordEncryptionMetrics.KROXYLICIOUS_ENCRYPTED_RECORDS_BASE_METER_NAME_TOTAL_METER_PROVIDER
```
Comment (ShubhamRwt) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java]: I planned to count the encrypted record here since it provides me the total record that are encrypted successfully but I can move this to `RecordEncryptionFilter` class itself. I was worried to initialize it in `RecordEncryptionFilter` class because I see that in the `maybeEncodeProduce()` method, it was possible that the Encryption failed somehow and I incremented the metrics even when the encryption failed

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java (hunk 3 above)
Comment (ShubhamRwt) [reply kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: Sorry I guess its copy/paste leftover

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+import static io.micrometer.core.instrument.Metrics.globalRegistry;
+
+public class RecordEncryptionMetrics {
+
+    public static final String VIRTUAL_CLUSTER_LABEL = "virtual_cluster";
+    public static final String TOPIC_NAME = "topic_name";
+
+    // Base Metric Names
+    protected static final String KROXYLICIOUS_ENCRYPTED_RECORDS_BASE_METER_NAME = "kroxylicious_encrypted_records_total";
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: I'm pretty sure that _total is generated by the framework.   Test it by doing a `curl http://localhost:<management port>`

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java (hunk 3 above)
Comment (ShubhamRwt) [reply kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: This PR needs a lot of cleaning

**Comment by ShubhamRwt** [comment]
This PR needs a lot of cleaning but I tried to express my thoughts with this

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    protected static final String KROXYLICIOUS_NOT_ENCRYPTED_RECORDS_BASE_METER_NAME = "kroxylicious_not_encrypted_records_total";
+
+    // Meter Providers
+    // (Callers use the providers to create the meters they need, augmenting with any tags).
+    public static final ClusterNodeSpecificMetricProviderCreator<Counter> KROXYLICIOUS_ENCRYPTED_RECORDS_BASE_METER_NAME_TOTAL_METER_PROVIDER = (virtualClusterName,
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: Take into account the comments @SamBarker and @tombentley are making on #2296 about this coding pattern.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java [hunk 7]
```diff
@@ -147,6 +156,10 @@ private CompletionStage<MemoryRecords> attemptEncrypt(@NonNull String topicName,
         });
     }
 
+    private void initMetrics(String virtualClusterName, String topicName) {
+
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java]: what's this for?

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/RecordEncryptionFilterIT.java [hunk 8]
```diff
@@ -668,6 +670,52 @@ void shouldGenerateOneDek(KafkaCluster cluster, Topic topic, TestKmsFacade<?, ?,
...
+                .endEndpoints()
+                .endManagement();
+
+        NamedFilterDefinition namedFilterDefinition = buildEncryptionFilterDefinition(testKmsFacade, UnresolvedKeyPolicy.PASSTHROUGH_UNENCRYPTED);
+        builder.addToFilterDefinitions(namedFilterDefinition);
+        builder.addToDefaultFilters(namedFilterDefinition.name());
+
+        try (var tester = kroxyliciousTester(builder);
+                var managementClient = tester.getManagementClient();
+                var producer = tester.producer(Map.of(ProducerConfig.LINGER_MS_CONFIG, 1000, ProducerConfig.BATCH_SIZE_CONFIG, 2));
```
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/RecordEncryptionFilterIT.java]: why the configuration?

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java (hunk 7 above)
Comment (ShubhamRwt) [reply kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/InBandEncryptionManager.java]: I left a comment on this, I plan to utilize it but haven't put enough mind into it. As I wanted to share my thoughts early, I left it here for now

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 9]
```diff
@@ -114,13 +115,18 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
...
+                    generateMetrics(context.getVirtualClusterName(), unresolvedTopicNames, topicNameToData);
+
                     var futures = kekSelection.topicNameToKekId().entrySet().stream().flatMap(e -> {
                         String topicName = e.getKey();
                         var kekId = e.getValue();
                         TopicProduceData tpd = topicNameToData.get(topicName);
                         return tpd.partitionData().stream().map(ppd -> {
                             MemoryRecords records = (MemoryRecords) ppd.records();
                             return encryptionManager.encrypt(
+                                    context.getVirtualClusterName(),
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: why did you choose to give the responsibility for ticking the metric to the EncryptionManager?    You want as fewer parts of the system as possible to know about Metrics - think Parnas and Information Hiding.

So you've already given REF knowledge for the metrics for one path (fine).. so now strive to keep to give it the responsibility for recording metrics for the other path too.

Hint - you'll want to chain onto EncryptionManager.encrypt() future :)

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import static io.micrometer.core.instrument.Metrics.globalRegistry;
+
+public class RecordEncryptionMetrics {
+
+    public static final String VIRTUAL_CLUSTER_LABEL = "virtual_cluster";
+    public static final String TOPIC_NAME = "topic_name";
+
+    // Base Metric Names
+    protected static final String KROXYLICIOUS_ENCRYPTED_RECORDS_BASE_METER_NAME = "kroxylicious_encrypted_records_total";
+    protected static final String KROXYLICIOUS_NOT_ENCRYPTED_RECORDS_BASE_METER_NAME = "kroxylicious_not_encrypted_records_total";
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: icky metric name! not_encrypted = unencrypted or plain.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java (hunk 5 above)
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: Think more broadly about metric naming conventions.  Filters need an approach to naming their filters to prevent chaos.   Something like: `kroxylicious_<filter type snake case>_encrypted_records_total`

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+    @NonNull
+    private static Meter.MeterProvider<Counter> buildCounterMeterProvider(String meterName,
+                                                                          String description,
+                                                                          String clusterName,
+                                                                          String topicName) {
+        return Counter
+                .builder(meterName)
+                .description(description)
+                // for RecordEncryptionFilterTest class, we need to set the cluster name as `test, else the test fails since cluster name tag cannot be null
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: we try to avoid leaking test concerns into production code.  Better to fix the test so that the virtual cluster has a name.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 12]
```diff
@@ -114,6 +115,10 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
                     if (!unresolvedTopicNames.isEmpty() && unresolvedKeyPolicy == UnresolvedKeyPolicy.REJECT) {
                         return CompletableFuture.failedFuture(new UnresolvedKeyException("failed to resolve key for: " + unresolvedTopicNames));
                     }
+
+                    // metrics generation for not encrypted record
+                    generateMetrics(context.getVirtualClusterName(), unresolvedTopicNames, topicNameToData);
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: can we give this method a better name?  If you get the method name right, the comment becomes superfluous.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 13]
```diff
@@ -139,6 +150,17 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
                 });
     }
 
+    private void generateMetrics(String virtualClusterName, Set<String> unresolvedTopicNames, Map<String, TopicProduceData> topicNameToData) {
+        unresolvedTopicNames.forEach(unresolvedTopic -> {
+            var unEncyptedRecordsTotal = RecordEncryptionMetrics.recordEncryptionPlainRecordsCounter(virtualClusterName, unresolvedTopic).withTags();
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: unencrypted is an English word, so you don't need the camel case.  But plain or plain text has the same meaning and is simpler.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 14]
```diff
@@ -139,6 +150,17 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
...
+            TopicProduceData data = topicNameToData.get(unresolvedTopic);
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: move the metric lookup into the loop.
you can then use unresolvedTopicNames.map(... /* to TopicProduceData */).filter(Objects::notNull).forEach(...)

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+public class RecordEncryptionMetrics {
+
+    public static final String VIRTUAL_CLUSTER_LABEL = "virtual_cluster";
+    public static final String TOPIC_NAME = "topic_name";
+
+    // Base Metric Names
+    private static final String RECORD_ENCRYPTION_ENCRYPTED_RECORDS = "kroxylicious_record_encryption_encrypted_records";
+    private static final String RECORD_ENCRYPTION_PLAIN_RECORDS = "kroxylicious_record_encryption_plain_records";
+
+    public static Meter.MeterProvider<Counter> recordEncryptionEncryptedRecordsCounter(String clusterName, String topicName) {
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: You are kind of misusing the MeterProvider API.  The use-case for it is where you know some labels in advance and others come later.

**Ideally** we would know the virtualCluster's name in the REF constructor so we could create the meter providers there.

```java
this.recordEncryptedCounters = RecordEncryptionMetrics                                               .recordEncryptionEncryptedRecordsCounter(context.getVirtualClusterName());
```

Inside `maybeEncodeProduce` we'd then have:

```java 
recordEncryptionPlainRecordsCounter.tags("topic", topicName) 
```

We can't do this at the moment.

I do think there is merit in initialising the MeterProviders with the virtual cluster at the begining of `maybeEncodeProduce` then specialising them with the topic_name later.

**Comment by k-wall** [comment]
We should have a unit test for metrics in RecordEncryptionFilterTest too.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 16]
```diff
@@ -126,7 +131,13 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
                                     new EncryptionScheme<>(kekId, EnumSet.of(RecordField.RECORD_VALUE)),
                                     records,
                                     context::createByteBufferOutputStream)
-                                    .thenApply(ppd::setRecords);
+                                    .thenApply(ppd::setRecords)
+                                    .thenApply(produceData -> {
+                                        var encyptedRecordsTotal = RecordEncryptionMetrics
+                                                .recordEncryptionEncryptedRecordsCounter(context.getVirtualClusterName(), topicName).withTags();
+                                        encyptedRecordsTotal.increment(RecordEncryptionUtil.totalRecordsInBatches((MemoryRecords) produceData.records()));
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]:
```suggestion
                                        RecordEncryptionMetrics
                                                .recordEncryptionEncryptedRecordsCounter(context.getVirtualClusterName(), topicName).withTags().increment(RecordEncryptionUtil.totalRecordsInBatches((MemoryRecords) produceData.records()));
```

Give or take the formatter?

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 17]
```diff
@@ -139,6 +150,17 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
                 });
     }
 
+    private void generateMetrics(String virtualClusterName, Set<String> unresolvedTopicNames, Map<String, TopicProduceData> topicNameToData) {
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: Generate what metrics? The collection of arguments doesn't seem to fit a generic metric generator function... Whats the difference between a metric generator and a MeterProvider?

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java (hunk 11 above)
Comment (SamBarker) [reply kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: Fix the test. We should not be hard coding test expectations in production code, far too many landmines that way.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+import static io.micrometer.core.instrument.Metrics.globalRegistry;
+
+public class RecordEncryptionMetrics {
+
+    public static final String VIRTUAL_CLUSTER_LABEL = "virtual_cluster";
+    public static final String TOPIC_NAME = "topic_name";
+
+    // Base Metric Names
+    private static final String RECORD_ENCRYPTION_ENCRYPTED_RECORDS = "kroxylicious_record_encryption_encrypted_records";
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: We are already in `RecordEncryptionMetrics ` what value does the `RECORD_ENCRYPTION_` prefix add to the constant name?

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    private static final String RECORD_ENCRYPTION_PLAIN_RECORDS = "kroxylicious_record_encryption_plain_records";
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: I think we should be pushing these metrics into a separate namespace from the core proxy runtime metrics. So 
```suggestion
    private static final String RECORD_ENCRYPTION_ENCRYPTED_RECORDS = "kroxylicious_filter_record_encryption_encrypted_records";
    private static final String RECORD_ENCRYPTION_PLAIN_RECORDS = "kroxylicious_filter_record_encryption_plain_records";
```

I'm a little wary of any filter using the `kroxylicious` prefix but I guess this is a project built and supported filter so its probably appropraite.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+    public static Meter.MeterProvider<Counter> recordEncryptionEncryptedRecordsCounter(String clusterName, String topicName) {
+        return buildCounterMeterProvider(RECORD_ENCRYPTION_ENCRYPTED_RECORDS, "Incremented by the number of records encrypted.",
+                clusterName, topicName);
+    }
+
+    public static Meter.MeterProvider<Counter> recordEncryptionPlainRecordsCounter(String clusterName, String topicName) {
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: again what is the `recordEncryption` prefix adding here?

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java (hunk 13 above)
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: consistency would be nice here. The records are either `plain` or `unencrypted` but both just begs the question "whats the difference between plain and unencrypted?"

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java (hunk 12 above)
Comment (SamBarker) [reply kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: My rule of thumb is comments should explain "why something is" rather than what it is, as that should be obvious from reading the code (with the obvious exception when dealing with some horribly named 3rd party API). They should generally be reserved for explaining subtle or easily over looked things.

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java (hunk 15 above)
Comment (ShubhamRwt) [reply kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]: I don't see how we can get the virtual cluster name in the REF constructor for now? The filter context is something present in the `maybeEncodeProduce` method itself. I can move the initialisation for them at the start of the method

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java (hunk 14 above)
Comment (ShubhamRwt) [reply kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: Sorry Keith, I tried this but wasn't able to get the desired result

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    private static final String ENCRYPTED_RECORDS = "kroxylicious_filter_record_encryption_encrypted_records";
+    private static final String PLAIN_RECORDS = "kroxylicious_filter_record_encryption_plain_records";
+
+    public static Meter.MeterProvider<Counter> encryptedRecordsCounter(String clusterName) {
+        return buildCounterMeterProvider(ENCRYPTED_RECORDS, "Incremented by the number of records encrypted.",
+                clusterName);
+    }
+
+    public static Meter.MeterProvider<Counter> plainRecordsCounter(String clusterName) {
+        return buildCounterMeterProvider(PLAIN_RECORDS, "Incremented by the number of records not encrypted.",
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]:
```suggestion
        return buildCounterMeterProvider(PLAIN_RECORDS, "A count of records the filter sent without encrypting them.",
```

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+    public static final String VIRTUAL_CLUSTER_LABEL = "virtual_cluster";
+    public static final String TOPIC_NAME = "topic_name";
+
+    // Base Metric Names
+    private static final String ENCRYPTED_RECORDS = "kroxylicious_filter_record_encryption_encrypted_records";
+    private static final String PLAIN_RECORDS = "kroxylicious_filter_record_encryption_plain_records";
+
+    public static Meter.MeterProvider<Counter> encryptedRecordsCounter(String clusterName) {
+        return buildCounterMeterProvider(ENCRYPTED_RECORDS, "Incremented by the number of records encrypted.",
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionMetrics.java]:
```suggestion
        return buildCounterMeterProvider(ENCRYPTED_RECORDS, "A count of the number of records which the filter encrypted.",
```

File: kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java [hunk 23]
```diff
@@ -139,6 +154,19 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
...
+                                             Map<String, TopicProduceData> topicNameToData) {
+        unresolvedTopicNames.forEach(unresolvedTopic -> {
+            TopicProduceData data = topicNameToData.get(unresolvedTopic);
+            if (data != null) {
+                data.partitionData()
+                        .forEach(produceData -> plainRecordsTotal
+                                .withTags(RecordEncryptionMetrics.TOPIC_NAME, unresolvedTopic)
+                                .increment(RecordEncryptionUtil.totalRecordsInBatches((MemoryRecords) produceData.records())));
+            }
+        });
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java]: In reply to https://github.com/kroxylicious/kroxylicious/pull/2308/files#r2152809984 How about:
```suggestion
                topicNameToData.entrySet().stream()
                .filter(topicDataEntry -> unresolvedTopicNames.contains(topicDataEntry.getKey()))
                .forEach(topicData -> topicData.getValue().partitionData()
                .forEach(produceData -> plainRecordsTotal
                        .withTags(RecordEncryptionMetrics.TOPIC_NAME, topicData.getKey())
                        .increment(RecordEncryptionUtil.totalRecordsInBatches((MemoryRecords) produceData.records()))));
```

File: kroxylicious-filters/kroxylicious-record-encryption/src/test/java/io/kroxylicious/filter/encryption/RecordEncryptionFilterTest.java [hunk 24]
```diff
@@ -169,6 +187,33 @@ void shouldNotEncryptTopicWithoutKeyId() {
...
+
+        // When
+        encryptionFilter.onProduceRequest(ProduceRequestData.HIGHEST_SUPPORTED_VERSION, new RequestHeaderData(), produceRequestData, context);
+
+        // Then
+        assertThat(Metrics.globalRegistry.get("kroxylicious_filter_record_encryption_plain_records").counter())
+                .isNotNull()
+                .satisfies(counter -> assertThat(counter.getId()).isNotNull())
+                .satisfies(counter -> assertThat(counter.count())
+                        .isGreaterThan(0));
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/test/java/io/kroxylicious/filter/encryption/RecordEncryptionFilterTest.java]: As this is a unit test we can be explicit about what to expect so we can do 
```suggestion
                        .isEqualTo(1.0));
```
Because we know its operating on exactly the records we expect and there is no background processing.

File: kroxylicious-filters/kroxylicious-record-encryption/src/test/java/io/kroxylicious/filter/encryption/RecordEncryptionFilterTest.java [hunk 25]
```diff
@@ -169,6 +187,33 @@ void shouldNotEncryptTopicWithoutKeyId() {
...
+
+        assertThat(Metrics.globalRegistry.get("kroxylicious_filter_record_encryption_encrypted_records").counter())
+                .isNotNull()
+                .satisfies(counter -> assertThat(counter.getId()).isNotNull())
+                .satisfies(counter -> assertThat(counter.count())
+                        .isGreaterThan(0));
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/test/java/io/kroxylicious/filter/encryption/RecordEncryptionFilterTest.java]: Similarly
```suggestion
                        .isEqualTo(1.0));
```

File: kroxylicious-filters/kroxylicious-record-encryption/src/test/java/io/kroxylicious/filter/encryption/RecordEncryptionFilterTest.java [hunk 26]
```diff
@@ -169,6 +187,33 @@ void shouldNotEncryptTopicWithoutKeyId() {
...
+        assertThat(Metrics.globalRegistry.get("kroxylicious_filter_record_encryption_plain_records").counter())
+                .isNotNull()
+                .satisfies(counter -> assertThat(counter.getId()).isNotNull())
+                .satisfies(counter -> assertThat(counter.count())
+                        .isGreaterThan(0));
+
+        assertThat(Metrics.globalRegistry.get("kroxylicious_filter_record_encryption_encrypted_records").counter())
+                .isNotNull()
+                .satisfies(counter -> assertThat(counter.getId()).isNotNull())
+                .satisfies(counter -> assertThat(counter.count())
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-record-encryption/src/test/java/io/kroxylicious/filter/encryption/RecordEncryptionFilterTest.java]: I'd be tempted to combine the two satisfies calls into one but happy to leave the choice up to you.

**Review by SamBarker — APPROVED** [review]
Thanks for all your efforts around this @ShubhamRwt I think your very close. 

I'd be happy to merge once the descriptions are updated.
